HISTORY.md

```diff
@@ -26,6 +26,7 @@
 * sst_dump to add a new --readahead_size argument. Users can specify read size when scanning the data. Sst_dump also tries to prefetch tail part of the SST files so usually some number of I/Os are saved there too.
 * Generate file checksum in SstFileWriter if Options.file_checksum_gen_factory is set. The checksum and checksum function name are stored in ExternalSstFileInfo after the sst file write is finished.
 * Add a value_size_soft_limit in read options which limits the cumulative value size of keys read in batches in MultiGet. Once the cumulative value size of found keys exceeds read_options.value_size_soft_limit, all the remaining keys are returned with status Abort without further finding their values. By default the value_size_soft_limit is std::numeric_limits<uint64_t>::max().
+* Enable SST file ingestion with file checksum information. Added files_checksum and files_checksum_func_name to IngestExternalFileArg such that use can call IngestExternalFiles with file checksum information. Add generate_and_verify_file_checksum to IngestExternalFileOptions (default is False). If it is set to True and DB enables file checksum, DB will generate the checksum for each ingested SST file and verify with the ingested checksum informaiton. Any mismatch will fail the ingestion. If the ingested checksum information is non-empty, DB enables file checksum, and generate_and_verify_file_checksum is False, DB will still verify the checksum function name. If the function name matches, the injested checksum will be directly stored in Manifest without verification. Note that, if IngestExternalFileOptions::write_global_seqno is True, the seqno will be changed in the ingested file. Therefore, the checksum of the file will be changed. In this case, no checksum information will be stored in the Manifest, unless generate_and_verify_file_checksum is set to True and a new checksum will be generated after the seqno is updated.
```

Comment:
How about `verify_file_checksum` instead of `generate_and_verify_file_checksum`? IMO "generate" implies the checksum originates in the `IngestExternalFiles()` layer, but I believe we may not need the option to control whether that happens (see below).

> Note that, if IngestExternalFileOptions::write_global_seqno is True, the seqno will be changed in the ingested file. Therefore, the checksum of the file will be changed. In this case, no checksum information will be stored in the Manifest, unless generate_and_verify_file_checksum is set to True and a new checksum will be generated after the seqno is updated.

I think for this case we can just generate it anyways when the DB enables file checksum, regardless of the value of `generate_and_verify_file_checksum`.

The other case `IngestExternalFiles()` may need to originate a checksum is when user did not provide a `files_checksum`. I think we can also use the DB setting to decide whether to generate a checksum here.

Reply:
My intent behind that suggestion is making it easier to explain. Here's the explanation after that change in my mind (there may be holes in it, please let me know): 

files_checksum contains the file checksum computed by the algorithm specified in files_checksum_func_name. The algorithm name is always checked for compatibility, and the checksum is optionally verified according to verify_files_checksum. Then files_checksum is written to the manifest. However, if files_checksum is not provided or not relevant due to write_global_seqno being set, we fall back to the DB options for the manifest's checksum.

Reply:
> How about `verify_file_checksum` instead of `generate_and_verify_file_checksum`

If I understand correctly: Whether the checksum of ingested file will be calculated is controlled by the DB checksum option. If DB enables the checksum, we will calculate it anyway. `verify_file_checksum` is used to decided if application want to verify whether the calculated checksum matches with the checksum provided by the application. Yeah, in this way the rule will be simpler.

Reply:
> Most of this seems like an API contract. HISTORY.md should describe the capabilities of the new feature, whether/how any existing API calls have changed behavior, and an API pointer in the form of names.

Will change the description in HISTORY.md after I switch to verify_file_checksum.